solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -2375,21 +2375,32 @@ public void rejoinShardLeaderElection(SolrParams params) {
 
   public void checkOverseerDesignate() {
     try {
-      byte[] data = zkClient.getData(ZkStateReader.ROLES, null, new Stat(), true);
-      if (data == null) return;
-      @SuppressWarnings({"rawtypes"})
-      Map roles = (Map) Utils.fromJSON(data);
-      if (roles == null) return;
-      @SuppressWarnings({"rawtypes"})
-      List nodeList = (List) roles.get("overseer");
-      if (nodeList == null) return;
-      if (nodeList.contains(getNodeName())) {
-        ZkNodeProps props = new ZkNodeProps(Overseer.QUEUE_OPERATION, CollectionParams.CollectionAction.ADDROLE.toString().toLowerCase(Locale.ROOT),
-            "node", getNodeName(),
-            "role", "overseer");
-        log.info("Going to add role {} ", props);
-        getOverseerCollectionQueue().offer(Utils.toJSON(props));
+      String s = System.getProperty("overseer.node");
+      Boolean isOverseer = s == null ? null : Boolean.parseBoolean(s);
+      byte[] data = null;
+      try {
+        data = zkClient.getData(ZkStateReader.ROLES, null, new Stat(), true);
+      } catch (NoNodeException e) {
+      }
+      @SuppressWarnings("unchecked")
+      Map<String, Object> roles = data == null ? Collections.emptyMap() : (Map<String, Object>) Utils.fromJSON(data);
+      List<String> overseerDesignates = (List<String>) roles.getOrDefault("overseer", Collections.emptyList());
+      CollectionParams.CollectionAction cmd = null;
+      if (isOverseer == null) {
+        if (overseerDesignates.contains(getNodeName())) cmd = CollectionParams.CollectionAction.ADDROLE;
+      } else {
+        if (isOverseer == Boolean.TRUE) {
+          cmd = CollectionParams.CollectionAction.ADDROLE;
+        } else if (isOverseer == Boolean.FALSE) {
+          if (overseerDesignates.contains(getNodeName())) cmd = CollectionParams.CollectionAction.REMOVEROLE;
+        }
       }
+      if (cmd == null) return;
+      ZkNodeProps props = new ZkNodeProps(Overseer.QUEUE_OPERATION, cmd.toLower(),
```

Comment:
Minor: if node is already in overseer role and we restart the node, then do we need to "addRole" again? (or it is just safe to add that role again?)